508 radius client failover

abujeda opened the pull request:
Added failover support to RADIUS client.

This PR is based on https://github.com/hmdc/sid2/pull/11. Merge PR 11 first before merging this one.

abujeda committed:
DevOpsProjects#483: Added skeleton project for Radius SPI

abujeda committed:
DevOpsProjects#483: Updated java package name to reflect iq.harvard.edu domain

abujeda committed:
feature(keycloak-radius-spi) GH #482 Added basic functionality for Keycloak Radius Authenticator

abujeda committed:
feature(keycloak-radius-spi) GH #482 Fixed build process by adding missing library

abujeda committed:
feature(keycloak-radius-spi) GH #482 Fixed missing library when running utility to check Radius connectivity

whorka committed:
Update README.md

Describe the configuration properties by key name, as used in Puppet.

abujeda committed:
feature(keycloak-radius-spi) GH #508 Added failover feature to Radius client

whorka commented on keycloak-radius-spi/Makefile:
```diff
@@ -0,0 +1,22 @@
+WORKING_DIR := $(shell pwd)
+
+build:
+	docker run -it --rm -v $(WORKING_DIR):/usr/local/app -w /usr/local/app maven:3-openjdk-11 mvn clean && mvn package
```
@adaybujeda can you help me to understand this?
- Why are you running `mvn clean` inside the `maven` container, but running `mvn package` in the host OS?
- What is the advantage to running the build inside the container if the host OS still requires Maven and Java?
- Couldn't it introduce problems if the version of Maven on the host OS is not the same as the one in the container?

abujeda replied:
That was my naive understanding on how shell commands work. I thought that `mvn clean && mvn package` were both running inside the container.

I have fixed the docker command so that both commands will run inside the container.
The idea was to have a consistent build environment with Docker and for Java and Maven to not be required.

whorka replied:
ok, that sounds good. I have updated the README.md to remove the requirements on java and maven there also.

whorka commented on keycloak-radius-spi/.gitignore:
```diff
@@ -0,0 +1,16 @@
+<<<<<<< Updated upstream
```
This looks like a merge artifact that needs to be removed

abujeda replied:
This is fixed now.

whorka commented:
The build failed when I ran the steps in the README.md:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  6.951 s
[INFO] Finished at: 2021-08-03T14:27:38-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project keycloak-radius-spi: Could not resolve dependencies for project edu.harvard.iq.keycloak:keycloak-radius-spi:jar:1.0: Could not find artifact org.tinyradius:tinyradius:jar:1.1.1 in central (https://repo.maven.apache.org/maven2) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
```

abujeda committed:
feature(keycloak-radius-spi) GH #508 Fixed build script + Updated Keycloak version libraries to 13.0.1

abujeda commented:
I have fixed the build script and updated the version of the Keycloak libraries to match the version of the server that we are using. I saw from the conversation Bill and Trey were having for the Puppet module that we are on `Keycloak 13.0.1`

abujeda committed:
feature(keycloak-radius-spi) GH #508 Improved build script by using a maven repo local to the project

whorka committed:
Update README.md

Update build requirements, build instructions, and Keycloak configuration instructions

whorka commented:
> I have fixed the build script and updated the version of the Keycloak libraries to match the version of the server that we are using. I saw from the conversation Bill and Trey were having for the Puppet module that we are on `Keycloak 13.0.1`

@adaybujeda could you please make this a configurable parameter, or at least describe in the README.md how to set the Keycloak version?

abujeda committed:
feature(keycloak-radius-spi) GH #508 Added documentation on how to update the Keycloak version for the Java libraries and local environment

abujeda commented:
@whorka I have added documentation to the README on how to update the Keycloak version.

As well, I have updated the version of the Keycloak Docker container to 13.0.1 (it was 14.0.0)

whorka commented:
I was able to build the jar locally, following the instructions in the README.md, and deploy it to the prototype Keycloak server. I then updated the config parameters by applying a new puppet manifest.

abujeda commented:
Already merged in from another PR
